### PR TITLE
Utilize DSIGNAlgorithm Ed25519Bip32DSIGN for extended keys in cardano-api

### PR DESCRIPTION
**Pull request by intricate**
The same as #2227, but reopened to bump it.

At the moment, Shelley extended signing keys are serialized using the `cardano-crypto` 128 byte binary format. However, these extended signing keys can actually be serialised to a more compact 96 byte format (see the `DSIGNAlgorithm Ed25519Bip32DSIGN` implementation).

The `DSIGNAlgorithm Ed25519Bip32DSIGN` implementation allows:

- Serialisation of BIP32-Ed25519 extended signing keys to a more compact 96 byte format as opposed to the old `cardano-crypto` 128 byte format.

- Deserialisation of BIP32-Ed25519 extended signing keys in either of the 96 or 128 byte formats.

### CODE REVIEW

**Comment by intricate** [comment]
bump @dcoutts @Jimbo4350

**Comment by intricate** [comment]
@newhoggy I would also note that I think this PR is still highly relevant

**Comment by Jimbo4350** [comment]
@intricate I added a test and rebased

**Review by newhoggy — APPROVED** [review]

**Review by carbolymer — APPROVED** [review]

**Comment by newhoggy** [comment]
Still needs further reviews.

**Comment by newhoggy** [comment]
Is this in any sense a breaking change?

**Comment by newhoggy** [comment]
Okay rebase done.

**Comment by newhoggy** [comment]
From what I can gather, these are the golden files of relevance to the PR:

```
test/data/golden/shelley/keys/converted_cardano-address_keys/byron_signing_key
test/data/golden/shelley/keys/converted_cardano-address_keys/icarus_signing_key
test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_payment_signing_key
test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_payment_signing_key_bip32
test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_payment_signing_key_legacy
test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_stake_signing_key
test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_stake_signing_key_bip32
```

File: cardano-cli/test/Test/Golden/Shelley/Key/ConvertCardanoAddressKey.hs [hunk 1]
```diff
@@ -186,7 +180,22 @@ golden_convertCardanoAddressShelleyStakeSigningKey =
...
+      liftIO . readFileTextEnvelope (AsSigningKey AsPaymentExtendedKey)
+        $ File "test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_payment_signing_key_bip32"
+    elegacy <-
+      liftIO . readFileTextEnvelope (AsSigningKey AsPaymentExtendedKey)
+        $ File "test/data/golden/shelley/keys/converted_cardano-address_keys/shelley_payment_signing_key_legacy"
+
+    bip32 <- evalEither ebip32
+    legacy <- evalEither elegacy
+
+    getVerificationKey bip32 === getVerificationKey legacy
```
Comment (newhoggy) [inline cardano-cli/test/Test/Golden/Shelley/Key/ConvertCardanoAddressKey.hs:201]: Right, so the legacy format still works.  That's good.

**Comment by intricate** [comment]
> Is this in any sense a breaking change?
> 
> For instance can it still deserialise the `128` byte format?
> 
> I'm thinking it might be useful to have a command that converts between these two byte formats as well.

This PR supports deserialisation in either of the 96 or 128 byte formats (see `xPrvFromBytes`).

I would probably vote against adding a command to convert between the two as the less compact 128 byte format isn't particularly useful. It should be fine to continue being able to deserialize it, but not serialize to it anymore.

**Comment by smelc** [comment]
Closing because of inactivity

**Pull request closed without merging**